filedistribution/src/main/java/com/yahoo/vespa/filedistribution/maintenance/FileDistributionCleanup.java

```diff
@@ -0,0 +1,109 @@
+                                                   int numberToAlwaysKeep,
+                                                   Set<String> fileReferencesInUse) {
+        log.log(Level.FINE, () -> "Keep unused file references for " + keepFileReferencesDuration);
+        if (!fileReferencesPath.isDirectory()) throw new RuntimeException(fileReferencesPath + " is not a directory");
+
+        log.log(Level.FINE, () -> "File references in use : " + fileReferencesInUse);
+
+        List<String> candidates = sortedUnusedFileReferences(fileReferencesPath, fileReferencesInUse, keepFileReferencesDuration);
+        // Do not delete the newest ones
+        List<String> fileReferencesToDelete = candidates.subList(0, Math.max(0, candidates.size() - numberToAlwaysKeep));
+        if (fileReferencesToDelete.size() > 0) {
+            log.log(Level.FINE, () -> "Will delete file references not in use: " + fileReferencesToDelete);
+            fileReferencesToDelete.forEach(fileReference -> {
+                File file = new File(fileReferencesPath, fileReference);
+                if (!IOUtils.recursiveDeleteDir(file))
+                    log.log(Level.WARNING, "Could not delete " + file.getAbsolutePath());
+            });
+        }
+        return fileReferencesToDelete;
+    }
+
+    private List<String> sortedUnusedFileReferences(File fileReferencesPath, Set<String> fileReferencesInUse, Duration keepFileReferences) {
+        Set<String> fileReferencesOnDisk = getFileReferencesOnDisk(fileReferencesPath);
+        log.log(Level.FINE, () -> "File references on disk (in " + fileReferencesPath + "): " + fileReferencesOnDisk);
+        Instant instant = clock.instant().minus(keepFileReferences);
+        return fileReferencesOnDisk
+                .stream()
+                .filter(fileReference -> !fileReferencesInUse.contains(fileReference))
+                .filter(fileReference -> isLastFileAccessBefore(new File(fileReferencesPath, fileReference), instant))
+                .sorted(Comparator.comparing(a -> lastAccessed(new File(fileReferencesPath, a))))
+                .collect(Collectors.toList());
+    }
+
+    private boolean isLastFileAccessBefore(File fileReference, Instant instant) {
+        return lastAccessed(fileReference).isBefore(instant);
+    }
+
+    private Instant lastAccessed(File fileReference) {
+        BasicFileAttributes fileAttributes;
+        try {
+            fileAttributes = readAttributes(fileReference.toPath(), BasicFileAttributes.class);
+            return fileAttributes.lastAccessTime().toInstant();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * Returns all files in the given directory, non-recursive.
+     */
+    public static Set<String> getFileReferencesOnDisk(File directory) {
+        Set<String> fileReferencesOnDisk = new HashSet<>();
+        File[] filesOnDisk = directory.listFiles();
```

Comment:
This has been known to cause OutOfMemoryError if the directory contains enough entries.  Consider reimplementing this, see UnixPath::listContentsOfDirectory.

Reply:
Yeah, will do something similar